Implement fake seek for VLC

QuantumEntangledAndy opened the pull request:
VLC seems to be peoples prefered media player

Went to look into why it plays so jittering on VLC but is ok in ffmpeg etc. Came back with these things:

- VLC is not a low latency player, you need to change some setting and it's still not perfect
- VLC does not seem to respect live NON SEEKABLE streams

To test things I printed the time stamps as being sent from neolink and watched the time stamps comming in on VLC
what was observed was that by the time VLC has finished starting up neolink is now pushing data about 5s ahead of
VLC's "now". VLC reports that buffers are "way too early" and starts adding zeros in.

What VLC seems to do is connect to the stream and then request a SEEK back to zero. Well neolink streams aren't seekable
resulting in the issues we see.

To address this I added a fake SEEK what it does is adjust the clients buffer start time such that new buffers are now at the requested seek time (even though really its just the same new buffers just stamped differently)

---

**Update:**

To address this even further additional changes have been made:

- maintain a buffer for local seeking. Rather that always updating to live on any seek request we now see if we can seek within a local buffer of about 3 IFrame + their associated Pframes and audio
- we now only push buffers onto the gatreamer pipeline if the clock matches the buffer. This really improves the connect speed of vlc and ffmpeg
-  <strike> If a client gets too far behind the stream then seek them back to live. This should help with keeping us closer to the live picture from the camera. </strike> Edit: Removed due to loop issues during stream errors

Possible related issues: #37 #38 #48 #56

Veuchez commented:
With this latest version I don't get any errors, I see much better, but I still have the video stream slightly jerky, both in mainstream and substream... the video stream is not yet as smooth as in Reolink streaming.

Veuchez commented:
With today's "version" we are almost perfect on VLC, congratulations. 
There are still a few small shots every now and then. 
On Frigate, on the other hand, I still see everything jerky.

QuantumEntangledAndy commented:
I see. Looks like I'm going to need to get a copy of frigate myself and see what's up.

My laptop however is in for repair for the week so will have to wait until after that before I can do any serious coding again.

Checksum commented:
Sorry been very busy at work. I did a bit of testing last week with this PR and did see some improvement. But reading your most recent fix and your comment on the VLC settings, I'll give it another go today.

Checksum commented:
The latest commit makes streaming MUCH better. The stream is now playable in VLC and doesn't skip much. After the initial delay of 3-5s buffering, VLC doesn't seem to drop that many frames.

Also, tested with [scrypted](https://github.com/koush/scrypted). My setup is Reolink Argus PT -> Neolink -> Scrypted -> Homekit and the stream is entirely usable. Previously I use to be able to see a max of 2 seconds every 15-20 seconds. This makes things a million times better.

Great work on figuring this out @QuantumEntangledAndy ! I'll keep an eye overnight to see, but I am almost certain looking at the logs that it should be fine over a period of time.

Checksum commented:
So had the neolink docker container running overnight. The stream doesn't work and I see the following in the logs:

```
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::factory] Building H264 Pipeline
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::factory] No clock
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 33.336ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 99.999ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 166.67ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 233.337ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 300.001ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 366.666ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 433.335ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::factory] Send seek H264: 0
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::factory]   - Sent seek H264: true
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 499.999ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - runtime: 112084.003670808s
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - buftime: 566.668ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 566.668ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - runtime: 112084.004924639s
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - buftime: 633.336ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 633.336ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - runtime: 112084.005280374s
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - buftime: 700.002ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 700.002ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - runtime: 112084.006116787s
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - buftime: 766.666ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 766.666ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - runtime: 112084.006300117s
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender]   - buftime: 833.335ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::sender] PTS: 833.335ms
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::factory] Send seek H264: 0
[2023-04-27T09:48:56Z DEBUG neolink::rtsp::gst::factory]   - Sent seek H264: true
```

Restarting the container fixes it. I imagine the `No clock` message is probably the most interesting?

QuantumEntangledAndy commented:
No clock is actually normal. The steam starts in a no clock state until it gets into the play state. It's the time when we send buffers for format identification rather than an actual stream. 

Is there anything else interesting in the stream?

Veuchez commented:
It happens to me too.  neolink works but the stream is not visible.

Checksum commented:
I just realised I didn't have the correct logging levels set for neolink 🤦🏽‍♂️ I'll give it another go today. In case it helps, this is the VLC log:

<img width="753" alt="image" src="https://user-images.githubusercontent.com/91416/234998946-319bc53b-81b3-4bc5-b2e4-93c12aec714f.png">

Checksum commented:
Some trace info from neolink. So the behaviour I noticed is that if I connect to neolink immediately after restarting, it works. But trying after about 10 mins doesn't succeed. I don't think it's got anything to do with the amount of time it waits for in between client requests.

https://gist.github.com/Checksum/2433b385d2830c16216fd92a73a0b5df

QuantumEntangledAndy commented:
The PTS log stops but we are still getting frames. It seems to happen after a jump to live. I'll have to check my logic there. But maybe will wait a week until I get my laptop back.

QuantumEntangledAndy commented:
Ive reworked my buffer code again. It should now handle jump to live better. Could you test once more?

Checksum commented:
> Ive reworked my buffer code again. It should now handle jump to live better. Could you test once more?

Thanks! Gave it a quick test. First time I connected with VLC, stream played with ~4s delay. Quality was steady. Then disconnect VLC and tried again, and no stream the second time around. Tried stopping/reconnecting VLC a few more times and noticed something bizzare: the stream only plays every _other_ time (1, 3, 5, ....). I found this weird, so restarted neolink and tried again and the same behaviour!

Didn't have access to my PC to capture logs, will try to do that tomorrow.

QuantumEntangledAndy commented:
Alright yet another build on the way. I think VLC was complaining about the buffers sent while in the preparing state. I've reworked those to only send prepration buffers up to the expected live point.

I am hoping that will do it for VLC. VLC seems to be very picky about its stream data. I will have a go at Frigate maybe it will also help. Does anyone have an example config? I'm trying to cobble one together from the docs but an actual in production one would help to make sure it works

Checksum commented:
> Alright yet another build on the way. I think VLC was complaining about the buffers sent while in the preparing state. I've reworked those to only send prepration buffers up to the expected live point.
> 
> I am hoping that will do it for VLC. VLC seems to be very picky about its stream data. I will have a go at Frigate maybe it will also help. Does anyone have an example config? I'm trying to cobble one together from the docs but an actual in production one would help to make sure it works

Just tried this out: panic and crashed unfortunately. Logs here - https://gist.github.com/Checksum/2433b385d2830c16216fd92a73a0b5df

QuantumEntangledAndy commented:
Ah this again. This happened on one of my tests but I couldn't replicate it. I need a full trace back to see which subtraction is failing.

QuantumEntangledAndy commented:
If you can get it reliably please add this environment variable RUST_BACKTRACE=1

Checksum commented:
> If you can get it reliably please add this environment variable RUST_BACKTRACE=1

Was trying that as I saw your comment :) Here is the stack trace:

```
neolink  | [2023-05-01T07:27:56Z DEBUG neolink::rtsp::gst::sender] Outside buffer jumping to live: 3.608896s < 5.655133567s < 5.608895s
neolink  | thread 'tokio-runtime-worker' panicked at 'overflow when subtracting durations', library/core/src/time.rs:936:31
neolink  | stack backtrace:
neolink  |    0: rust_begin_unwind
neolink  |              at ./rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:579:5
neolink  |    1: core::panicking::panic_fmt
neolink  |              at ./rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:64:14
neolink  |    2: core::panicking::panic_display
neolink  |              at ./rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:147:5
neolink  |    3: core::panicking::panic_str
neolink  |              at ./rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:131:5
neolink  |    4: core::option::expect_failed
neolink  |              at ./rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/option.rs:2045:5
neolink  |    5: core::option::Option<T>::expect
neolink  |              at ./rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/option.rs:905:21
neolink  |    6: <core::time::Duration as core::ops::arith::Sub>::sub
neolink  |              at ./rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/time.rs:936:31
neolink  |    7: neolink::rtsp::gst::sender::NeoMediaSender::jump_to_live::{{closure}}
neolink  |    8: neolink::rtsp::gst::sender::NeoMediaSenders::handle_buffer::{{closure}}
neolink  |    9: <neolink::rtsp::gst::factory::NeoMediaFactoryImpl as core::default::Default>::default::{{closure}}
neolink  |   10: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
neolink  |   11: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
neolink  |   12: tokio::runtime::task::harness::Harness<T,S>::poll
neolink  |   13: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
neolink  |   14: tokio::runtime::scheduler::multi_thread::worker::Context::run
neolink  |   15: tokio::macros::scoped_tls::ScopedKey<T>::set
neolink  |   16: tokio::runtime::scheduler::multi_thread::worker::run
neolink  |   17: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
neolink  |   18: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
neolink  |   19: tokio::runtime::task::harness::Harness<T,S>::poll
neolink  |   20: tokio::runtime::blocking::pool::Inner::run
neolink  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
neolink  | Error: task 5 panicked
```

QuantumEntangledAndy commented:
Thanks next build should fix this

QuantumEntangledAndy commented:
Ok should work now, I had a few short tests in frigate too and everything seemed ok in the logs

Checksum commented:
Can't see the new docker tag - looks like a few CI steps failed. Will try again when the image is published.

QuantumEntangledAndy commented:
Yes the GitHub runners seem to be having internet connectivity issues. It keeps failing at the steps where it downloads the dependencies. Usually just needs a rerun.

QuantumEntangledAndy commented:
Ok built

```bash
docker pull quantumentangledandy/neolink:fix-vlc_seek
```

Checksum commented:
Can't view the stream in VLC :( Logs attached.

[neolink.log](https://github.com/QuantumEntangledAndy/neolink/files/11369354/neolink.log)

QuantumEntangledAndy commented:
Yeah just fixed that

QuantumEntangledAndy commented:
Should be ready again 

```bash
docker pull quantumentangledandy/neolink:fix-vlc_seek
```

QuantumEntangledAndy commented:
When a stream is not yet ready (camera connecting) neolink puts up a dummy stream with static and the message stream not ready.

I noticed that frigate would not refresh that stream automatically once the real stream becomes avaliable so I added something in neolink to drop all current sessions when the stream is avaliable.

This seems to be working to force frigate to reconnect to the real stream.

Checksum commented:
No stream at all in VLC unfortunately.

QuantumEntangledAndy commented:
What sort of issue is it? Any logs or anything? I can confirm it runs on metal with me vlc etc

Checksum commented:
Just don't see any video window in VLC. Not even a black video. Have attached both logs

[VLC Debug Log (3.0.18 Vetinari).txt](https://github.com/QuantumEntangledAndy/neolink/files/11370029/VLC.Debug.Log.3.0.18.Vetinari.txt)
[neolink.log](https://github.com/QuantumEntangledAndy/neolink/files/11369998/neolink.log)

QuantumEntangledAndy commented:
Seems to be an issue with the docker. It keeps reporting media not prepared for me though docker (but same code base works on metal). This may mean that the gstreamer needs a version bump. I'll try and build the docker image locally and maybe update it to bullseye

QuantumEntangledAndy commented:
Your log is a bit different then I expected. Do you usually have lines along the way of

`: 3.608899s < 9.86991296s < 5.608897s`

The first and last of these are the time stamps from the camera. Which are low since mine usually start at 10000 which is what is stopping us from jumping to live as we cannot set the start_time to a negative value.

Also is there anything in your full log about recieving I/PFrames? I cannot seem to see any of those

QuantumEntangledAndy commented:
Whats happening here:

`: 3.608899s < 9.86991296s < 5.608897s`

is that the camera has given us data from 3s to 5s but we want to play at 9s. But there is no data at 9s. The current solution is to shift the start time backwards such that when asking for data at 9s we actually get data at 4s.

But this fails when the buffers starts so close to zero because of the use of unsigned integers (no negatives allowed).

Next build should address this by allowing the start time to go negative

QuantumEntangledAndy commented:
When you have the time another test would be appreciated. It is working for me now in docker on vlc so hopefully for you too

Veuchez commented:
Both in VLC on Windows and in frigate I see the streaming all jerky...

QuantumEntangledAndy commented:
What are your settings for jitter buffer, clock synchronisation and network cache on vlc?

Veuchez commented:
I have the default settings

QuantumEntangledAndy commented:
I see. I cannot replicate this. Are you getting lots of messages about jumping to live in the debug logs? Perhaps you could post them. 

The only time I get any issues any more is when the stream from the camera is not solid enough. But this is a seperate issue about connectivity not related to VLC buffer part of the code.

QuantumEntangledAndy commented:
Set the environmental variable RUST_LOG=debug

I'm not saying general connectivity issues. But issues perhaps in the other part of the neolink code base. Things like not pulling frames off of the network quickly enough or errors in decoding the frames. There's a lot of moving parts to get things working.

It would be helpful for testing this vlc stuff to do so on the subStream over tcp so we can eliminate that as a source of issue.

Once I can get the vlc working when the packets are coming in well I can look into improving things where the packets are coming in.

I know for example that sometimes the camera drops us when using udp (still haven't figured out why) and after such an event we have to do a reconnect and that causes issues with the playback.

Veuchez commented:
How do I save the log? there are so many lines.

Checksum commented:
I see some really weird playback issue as well. VLC starts out with a blank video for a couple of seconds (probably buffering), and then shows the live stream. But the stream is caught in a loop. The timestamp in the video shows that its just stuck at the same second and keeps looping. 

<img width="882" alt="Screenshot 2023-05-03 at 7 01 19 am" src="https://user-images.githubusercontent.com/91416/235785474-88be4120-3bf5-456c-bd1a-a87a7413931f.png">

[neolink.log](https://github.com/QuantumEntangledAndy/neolink/files/11377757/neolink.log)

QuantumEntangledAndy commented:
A loop happens when frames are not comming in from the camera any more.

Let me try to explain

- camera sends frames
- we get say 5s of data
- we store 2s in our buffer
This means we have frame data
From 3-5s
- vlc starts playing we mark its start time as 4s
- vlc plays for 1s
- then vlc wants frames from 5-6s which we don't have
Instead the start time is reset to 4s

I'll have a run though my code again and see if there's a cleaner way to do this but vlc seems quite picky in how it wants its streams.

If I stop sending frames vlc will drop us

If i send the last frame again vlc will say frame too late and still drop us

QuantumEntangledAndy commented:
Ok so removed the jump to live bit of the code. This means reduced error recovery capabilities. We now just leave it to VLC to drop us if it wants. While ffmpeg will just wait forever.

Also swapped to a non-blocking gstreamer queue, this might have been why we stopped getting new frames if we blocked the threads the code wouldn't be able to pull frames from the camera anymore

QuantumEntangledAndy commented:
Took a bit more work to get this one ready.

Changes
- No jump to live, just leave it on the client
- Use non-blocking gstreamer queues
- Close all clients when buffers are prepared
- Also prepare buffers once
- Use multi-threaded tokio runtim
- Various changes to speed up github actions
- Added some tokio task debugging code

If anyone has time to test please do

I am currently running VLC for 5mins now with neolink in docker with no issues. Will keep running and see

Veuchez commented:
I don't know if I'm testing the right version, I downloaded it from here.
https://github.com/QuantumEntangledAndy/neolink/suites/12643658589/artifacts/677373974

I keep seeing everything jerky in both mainstream and substream.
In VLC it looks better, on FRIGATE it goes badly.

There are many freezes and video artifacts.
What program can I use on Windows to save you a clear log?

Checksum commented:
No luck yet unfortunately. VLC doesn't seem to be picking up the stream at all. Also, if I stop and play again, VLC can't connect to the server. Also tried with Scrypted which uses ffmpeg, and same behaviour there as well. 

![image](https://user-images.githubusercontent.com/91416/236073376-3a01e1c1-26e5-48bc-8fa4-740b40ce64dc.png)

QuantumEntangledAndy commented:
@Veuchez  Last set of builds is here https://github.com/QuantumEntangledAndy/neolink/actions/runs/4871273443

@Checksum that error seems to suggest not even being able to connect to the rtsp server at all. That usually happens if there is something up with your config. If your on docker please ensure you are forwarding 8554 and that you have the bind address in the config as 0.0.0.0

Checksum commented:
> @Veuchez Last set of builds is here https://github.com/QuantumEntangledAndy/neolink/actions/runs/4871273443
> 
> @Checksum that error seems to suggest not even being able to connect to the rtsp server at all. That usually happens if there is something up with your config. If your on docker please ensure you are forwarding 8554 and that you have the bind address in the config as 0.0.0.0

@QuantumEntangledAndy they look correct because I'm able to connect for the _first_ time after starting the server. Subsequent connections are failing.

QuantumEntangledAndy commented:
@Checksum Could you test with just one camera in the config and for that camera config to have the setting `stream = "subStream"`

I notice that your log stops receiving frames after your connection. I want to eliminate bandwidth as an issue here

Checksum commented:
So have only 1 Reolink cam and have always used the subStream :-) I just figured what was wrong - I had the pause config on which was causing all the trouble. Without the pause config, the stream works really well. There is a delay of about 3s, but that's not too bad. 

So takeaway: change looks good, smoother than before, but pause config broken.

QuantumEntangledAndy commented:
Alright. That's a good sign at least. I was starting to tear out my hair with all this non reproducible trouble.

Pausing shuts down the stream which means the buffers are all old by the time the new data arrives. I think I'll need to work something out for this.

Checksum commented:
> Alright. That's a good sign at least. I was starting to tear out my hair with all this non reproducible trouble.
> 
> Pausing shuts down the stream which means the buffers are all old by the time the new data arrives. I think I'll need to work something out for this.

So just curious as to how it works now without the pause: does the initial connection time from neolink to the camera being used as the “start” time stamp is some way and when a client connects, is that used to jump to the current time?

Veuchez commented:
I still don't see smooth streaming with Argus 3 Pro.
On VLC it's better but the count of seconds still doesn't run smoothly... a second sometimes lasts two seconds, sometimes half a second.
On FRIGATE the streaming is bad and the seconds count goes every 2 seconds.

Veuchez commented:
Sometimes I get these errors, the stream stops on both VLC and FRIGATE and the stream is no longer reachable unless restarting neolink.

```
[2023-05-04T08:10:31Z INFO  neolink_core::bc_protocol] ingresso: Trying TCP discovery
[2023-05-04T08:10:36Z INFO  neolink_core::bc_protocol] ingresso: Trying local discovery
[2023-05-04T08:10:38Z INFO  neolink_core::bc_protocol] ingresso: Local discovery success 95270002XXXXXX4 at 192.168.2.101:35406
[2023-05-04T08:10:39Z INFO  neolink::rtsp::states::loggedin] ingresso: Camera time is already set: 2023-05-04 10:10:41.0 -01:00:00
[2023-05-04T08:10:39Z INFO  neolink::rtsp::states::loggedin] ingresso: Camera reports firmware version v3.0.0.907_22042201
[2023-05-04T08:10:39Z INFO  neolink::rtsp::states::streaming] ingresso: Starting video stream Main Stream (Clear)
[2023-05-04T08:10:39Z INFO  neolink::rtsp::states::streaming] ingresso: Starting video stream Sub Stream (Fluent)
[2023-05-04T08:40:53Z INFO  neolink::rtsp] Join Pause
[2023-05-04T08:40:53Z WARN  neolink::rtsp] ingresso: Retryable error: ingresso: Error while streaming

    Caused by:
        deadline has elapsed
[2023-05-04T08:40:53Z INFO  neolink_core::bc_protocol] ingresso: Trying TCP discovery
[2023-05-04T08:40:58Z INFO  neolink_core::bc_protocol] ingresso: Trying local discovery
[2023-05-04T08:41:00Z INFO  neolink_core::bc_protocol] ingresso: Local discovery success 95270002XXXXXXX4 at 192.168.2.101:35406
[2023-05-04T08:41:01Z INFO  neolink::rtsp::states::loggedin] ingresso: Camera time is already set: 2023-05-04 10:41:03.0 -01:00:00
[2023-05-04T08:41:01Z INFO  neolink::rtsp::states::loggedin] ingresso: Camera reports firmware version v3.0.0.907_22042201
[2023-05-04T08:41:01Z INFO  neolink::rtsp::states::streaming] ingresso: Starting video stream Main Stream (Clear)
[2023-05-04T08:41:01Z INFO  neolink::rtsp::states::streaming] ingresso: Starting video stream Sub Stream (Fluent)
```

QuantumEntangledAndy commented:
Deadline has elapsed means that neolink has not received a frame in 15s. It's something I'm looking into. But not likely related to this vlc compatability.

QuantumEntangledAndy commented:
Can you show me the log during the time when the frames are non uniform. I have an idea as to why that might happen but want to confirm it with the log

Veuchez commented:
If you can tell me how to save the registry with windows gladly. The rows are so many and they don't remain in memory...

QuantumEntangledAndy commented:
Oh you have to pipe the output to file.

Same for all terminal commands:

```cmd
SomeCommand > log.txt 2>&1
```

QuantumEntangledAndy commented:
This will redirect all output so you won't see anything on the terminal anymore just inside the file.

Veuchez commented:
Perfect thanks!
Here is the log.
[log.zip](https://github.com/QuantumEntangledAndy/neolink/files/11397105/log.zip)

QuantumEntangledAndy commented:
Ok when it finishes building I'd appreciate a test on the build. This one include a very slow (2.5s) smoothing that attempts to bring the live time to within the buffer (at 75% of the buffer). The effect is mostly not noticable accept when there is high load, but eventually it should even back out. This should keep us within the buffered zone so that we can deliver frames to VLC without stutter.

QuantumEntangledAndy commented:
Buikd is [here](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4889814406)

Veuchez commented:
Definitely much much better but I still see jerky.
Seems like the right way.
Here is the new log.
[log.txt](https://github.com/QuantumEntangledAndy/neolink/files/11404077/log.txt)

QuantumEntangledAndy commented:
Just has another idea about how to process the buffers more often. Should help a little bit more. Please check again when possible [build here](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4893830202)

Veuchez commented:
Even better, the stream is always smoother but there is still some stutter in both video and audio.
VLC keeps crashing.
Here is the log.
[log.txt](https://github.com/QuantumEntangledAndy/neolink/files/11406854/log.txt)

Veuchez commented:
The level reached is good, there are moments where streaming is perfect.
I don't understand why after a while the stream is stopped the only way to see the stream again is to restart neolink.

QuantumEntangledAndy commented:
I went to fix your bug about not restarting after awhile. It was from how the camera times are reset to zero after a reconnect so the timestamps in the buffer are all wrong.

I fixed this by resetting the buffer on a reconnect.

While there I had an idea about the jitter, previously frames were popped from the buffer in sets (of iframes) this would mean that the target buffer position would at time jump. The spring would then kick in to cause stutter

To fix this I changed how frames are popped AND how the target time is computed. 

I've been running it on my end for awhile with smooth 1s latency in ffmpeg and 3s latency in vlc (vlc has more buffer on its end)

Please could you give it a test. Specifically I need a test over reconnect events and long term running

QuantumEntangledAndy commented:
Build is [here](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4904876801)

Checksum commented:
Did about an hour's testing. Stream and stutter is definitely better for ffmpeg (scrypted). Client reconnects work flawlessly as well. However, seeing a strange stutter with VLC which skips a second, and then catches up. Video clip attached.

Thanks for all your hard work, just following the commits in this PR explains how tricky it is to make it play well with both VLC and other clients. Unfortunately I can't use neolink in real life as the pause function is broken and my Argus PT charges over a solar panel which isn't doing a really good job in keeping it charged. Happy to test out new builds and provide feedback in any way I can.

https://user-images.githubusercontent.com/91416/236659280-5a37b7fd-7873-47fe-a160-0f6920c669c0.mov

Veuchez commented:
Here's what I've noticed with this version.
So far I had the bitrate 2048kbps (3072 is the default for the Argus 3 PRO).
With this version the video streaming seems perfect to me!
I understand that there is some interruption only because of the audio that occasionally freezes.
VLC never crashed, it makes some video artifacts that you don't see in FRIGATE.
Towards the end of the log I posted I brought the bitrate to 3072kbps and here it seems that the buffer isn't behind it.
Streaming is still smooth but VLC crashes.
On FRIGATE, however, the streaming does not stop but stops to wait for the flow which resumes after a while.
Good job!!
[log.zip](https://github.com/QuantumEntangledAndy/neolink/files/11414230/log.zip)

QuantumEntangledAndy commented:
Yes it does seem better. I have run the stream for over 6h on ffmpeg. VLC lasted about about 2 hours.

From watching the log VLC disconnects when there are no frames being sent (during a reconnect), while ffmpeg will wait. There is probably a config option to configure the timeout in VLC but I think it is working well enough

Veuchez commented:
It seems that way to me as well. 
Maybe it seems slightly slow when I see cars passing by, but it could be just my imagination. 
Is there anything that can be done to improve the streaming with the 3072 bitrate? 
And what about the audio interruptions?

QuantumEntangledAndy commented:
@Checksum See if the next build fixes the client pause

Veuchez commented:
I confirm the slight slow motion; I saw my dog jump and it looked like it was in slow motion.
In a little while I will try the new version.

Veuchez commented:
The 3072 stream is slightly better, VLC doesn't freeze, but the audio and video still sometimes cut out. The slow motion effect is still present

QuantumEntangledAndy commented:
For the slow motion can you try this:

Put a system clock next to your camera feed. Once stable record the time difference. Wait an hour then record the difference again. 

I've already done this for on ffmpeg and vlc over the 6 hour period. Ffmpeg was only 1s behind the entire time and vlc only 3-4s behind. If there is a slow motion then over an hour it should add up to a very noticeable time difference between the two.

If however it could also be periods of slow and periods of fast. Which given the code I expect to happen from time to time as the stream is speed up and slowed down to achieve constant latency. You can think of the stream as on average being at the right rate but over short periods you can see some variance.

QuantumEntangledAndy commented:
The audio is more problematic. Reolink in their infinite wisdom do not timestamp the audio at all. Currently what I do is use the last time stamp from the nearest video frame but this can cause issue.

Veuchez commented:
I left neolink at 3072 kbps for a few hours, and when I got back home, the streaming was not available on either FRIGATE or VLC. 
[log.zip](https://github.com/QuantumEntangledAndy/neolink/files/11414932/log.zip)
I'm going back to 2048 and trying again. 
Then, I will also do the clock test.

Veuchez commented:
I am doing the clock test with a bitrate of 2048kbps on VLC. 
The seconds are really strange. After 15 minutes, the time in the video was 8 seconds behind, shortly after it was 10 seconds ahead, then it returned to the correct time and started accumulating delay, slowing down. 
After 20 minutes, the streaming stopped and I had to restart. 
However, the behavior is always the same, it accumulates delay compared to the clock, then recovers by accelerating, then slows down again, and so on continuously. 
Is there a way to avoid these accelerations and slowdowns and have a constant flow?

Veuchez commented:
Another thing, I noticed that with the latest version I couldn't see the live streaming on FRIGATE while with the previous one I could

Checksum commented:
> @Checksum See if the next [build](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4906317664) fixes the client pause

No luck with the pause config on yet. Didn't even bother testing on VLC cause the stream didn't start on ffmpeg. Logs attached.

[neolink.log](https://github.com/QuantumEntangledAndy/neolink/files/11417917/neolink.log)

QuantumEntangledAndy commented:
> > @Checksum See if the next [build](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4906317664) fixes the client pause
> 
> No luck with the pause config on yet. Didn't even bother testing on VLC cause the stream didn't start on ffmpeg. Logs attached.
> 
> [neolink.log](https://github.com/QuantumEntangledAndy/neolink/files/11417917/neolink.log)

Your using both client and motion pause, can you test one at a time. Just client seems to work for me

QuantumEntangledAndy commented:
I have another build which rewrites it to something a liittle simpler. Seems to be working for me, audio and clock rate, but as always your are probably going to tell me it's not working for you :)

QuantumEntangledAndy commented:
Just pushed an update that gets pausing working when both client and motion pause are on. All works fine in ffmpeg, client pause works fine in vlc but motion pause and vlc don't get on because theres no buffer comming in during motion pause and vlc drops it (ffmpeg will wait)

Build is [here](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4913062915)

Checksum commented:
> Just pushed an update that gets pausing working when both client and motion pause are on. All works fine in ffmpeg, client pause works fine in vlc but motion pause and vlc don't get on because theres no buffer comming in during motion pause and vlc drops it (ffmpeg will patiently wait for data)
> 
> Build is [here](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4913062915)

This build is looking _much_ better with client pause. ffmpeg is definitely more forgiving - I see it starts a bit behind, but catches up with a big jump in the timestamp (~7-8s). Stream is pretty stable and reconnects work well.

However, when I turn motion pause on, I only see a single frame (static image) from the stream. No updates at all.

> I have another [build](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4912366922) which rewrites it to something a liittle simpler. Seems to be working for me, audio and clock rate, but as always your are probably going to tell me it's not working for you :)

Yea, sorry to always bring in the bad news! Someday when I start playing with Rust, I can actually help out rather than just complain :)

QuantumEntangledAndy commented:
When you have motion pause on do you see messages about pausing and resuming in the log?

QuantumEntangledAndy commented:
Another new build. I noticed that the audio in VLC would be play->pause->play->pause. I had an idea that this was because we pushed two+ audio buffers with the same time stamp (until a new vid frame comes along with a new stamp). 

So to address this I now group audio buffers inbetween video frames so they are all part of the same buffer with only one timestamp.

This seems to help as far as I can tell.

QuantumEntangledAndy commented:
Build is [here](https://github.com/QuantumEntangledAndy/neolink/actions/runs/4915035427)

QuantumEntangledAndy commented:
Ok so I've tested the latest in ffmpeg, vlc and frigate. I think I am going to merge. But I'll wait for a bit in case your going to tell me it's still not working for you. If you want to test please do.

The only current issue I have is the `Timed out waiting for new Media Frame` which will need fixing in other bits of the code base

Veuchez commented:
I'm trying the latest version...
I can not understand, but the "slow motion" effect is only me? Maybe it depends on my hardware configuration or something.
In previous versions, however, this effect seemed much less evident, and the streaming seemed much more fluid, now I see it jerky again..

QuantumEntangledAndy commented:
As long as data is comming in ok. I don't see anything but a good stream. I do get the occasional data loss, which results in slow motion then an eventual catch up to live which can seem jerky.

However provided there are no error in the stream all is good.

Perhaps you can watch the debug log as you play the stream. You should notice that the issues occur at the same time as the debug log prints one of the error kinds. Although sometimes it take a good 15s before the error message appears because it needs to wait for the 15s timeout.

Veuchez commented:
I see the streaming constantly in slow motion in this version... in the next few days I'll have the opportunity to test it better anyway, then I'll also try it on Linux, maybe it's better than on Windows...

QuantumEntangledAndy commented:
Next build disables smoothing, I also added some logs to show the buffer size, anything less than 60 means the buffer is running behind

QuantumEntangledAndy commented:
Could you test this build with the count on the buffer size. It would let me confirm if it is because of the camera not filling up the buffer.

Veuchez commented:
Here is the log.
[log.txt](https://github.com/QuantumEntangledAndy/neolink/files/11448413/log.txt)

QuantumEntangledAndy commented:
Thanks for the log it confirms that the buffer is exhausted. There are not enough frames coming from the camera to maintain the fps

Veuchez commented:
Is it fixable?

QuantumEntangledAndy commented:
Maybe. The official client may have a way to dynamically reduce the fps when this starts happening to keep the buffer fill. Id need to observe that. You could also use the official client to change the resolution or fps.

QuantumEntangledAndy commented:
Once i get the time I will try a few things to improve the rate that packets are recieved, see if I can find the bottle neck. If it is to do with the UDP layer then I might be able to do something and make that more efficent, if it turns out we are network bound then I might be able to reduce/increase fps dynamically to keep the buffer filled.

For now though I think I will merge this. As long as the data comes in then it seems that the stream on ffmpeg/vlc/frigate is working and that was what I was aiming to achieve with this PR
